packages/query/src/block-processor.ts

```diff
@@ -103,89 +115,126 @@ export class BlockProcessor implements BlockProcessorInterface {
       // - decrypts new swaps
       // - updates idb with advice
       //const shouldStoreProgress = await this.viewServer.scanBlock(compactBlock);
-      const wasmWantsFlush = await this.viewServer.scanBlock(compactBlock);
+      const scannerWantsFlush = await this.viewServer.scanBlock(compactBlock);
 
       // flushing is slow, avoid it until
       // - wasm says
       // - every 1000th block
       // - every block at tip
       const flushReasons = {
-        wasmWantsFlush,
+        scannerWantsFlush,
         interval: compactBlock.height % 1000n === 0n,
         new: compactBlock.height > latestBlockHeight,
       };
 
       if (flushReasons.new) console.log('finished sync', performance.measure('start'));
 
-      let compactScan;
+      // state commitments identifying new records in this compact block, with a
+      // boolean to indicate if we should recover their source id
+      const blockCommitments = new Map<CommitmentHexString, NeedsSourceId>();
+      let flush;
       if (Object.values(flushReasons).some(Boolean)) {
-        compactScan = this.viewServer.flushUpdates();
-        console.log('flush', compactBlock, flushReasons, compactScan);
+        flush = this.viewServer.flushUpdates();
+        console.log('flush', compactBlock, flushReasons, flush);
 
         // in an atomic query, this
         // - saves 'sctUpdates'
         // - saves new decrypted notes
         // - saves new decrypted swaps
         // - updates last block synced
-        await this.indexedDb.saveScanResult(compactScan);
+        await this.indexedDb.saveScanResult(flush);
 
         // - detect unknown asset types, query the chain for asset metadata
         // - or, generate default fallback metadata
         // - update idb
         // this makes a network query.
         // TODO: should this be scanning swaps as well?
-        await this.identifyNewAssets(compactScan.newNotes);
+        await this.identifyNewAssets(flush.newNotes);
+
+        // boolean indicates if we should recover source id
+        for (const { noteCommitment, source } of flush.newNotes)
+          blockCommitments.set(
+            uint8ArrayToHex(noteCommitment!.inner),
+            source!.equals(blankTxSource),
+          );
+        for (const { swapCommitment, source } of flush.newSwaps)
+          blockCommitments.set(
+            uint8ArrayToHex(swapCommitment!.inner),
+            source!.equals(blankTxSource),
+          );
```

Comment:
this seems completely equivalent to compactBlock.sctUpdates.state_commitments

Reply:
@Valentine1898 any insight to this? will a list generated this way, and the contents of sctUpdates.state_commitments ever differ?

Reply:
The idea is that we don't need to create a 
`Map<CommitmentHexString, NeedsSourceId>()` , we just need to use `newNotes` and `newSwaps` when processing transactions and redefine the source for them. 
 There is also no reason for `resolveNullifiers` to return `Map<CommitmentHexString, NeedsSourceId>()`. When we spend a `nullifier` we should never override `source` (even if it suddenly turns out to be undefined). If we find a spent nullifier in a transaction we should only save that transaction.

Reply:
https://github.com/penumbra-zone/penumbra/blob/2b0408879aaef969aade53cb93cff1dad69b99ce/crates/view/src/worker.rs#L141 

This code on rust (lines 141-175) does everything we do, but simpler. At the same time we have all the same data as rust:
`new_notes`
`new_swap`
`spent_nullifiers` ( it is assumed that resolveNullifiers() will just return a list of nullifiers)